File: src/utils/openAIService.ts
import { SourceDocument } from './agentTypes';

// Define the ChatCompletionMessageParam type locally since we can't import it from openai/resources
interface ChatCompletionMessageParam {
  role: "system" | "user" | "assistant" | "function" | "tool";
  content: string;
  name?: string;
}

// This class handles communication with OpenAI API
export class OpenAIService {
  private apiKey: string | null = null;

  constructor(apiKey?: string) {
    if (apiKey) {
      this.apiKey = apiKey;
    }
  }

  setApiKey(apiKey: string) {
    this.apiKey = apiKey;
  }

  hasApiKey(): boolean {
    return !!this.apiKey;
  }

  // Generate embeddings for text using OpenAI's API
  async generateEmbeddings(texts: string[]): Promise<number[][]> {
    if (!this.apiKey) {
      throw new Error("OpenAI API key is not set");
    }

    try {
      const response = await fetch("https://api.openai.com/v1/embeddings", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Authorization": `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          input: texts,
          model: "text-embedding-3-small"
        })
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`OpenAI API error: ${JSON.stringify(error)}`);
      }

      const data = await response.json();
      return data.data.map((item: any) => item.embedding);
    } catch (error) {
      console.error("Error generating embeddings:", error);
      throw error;
    }
  }

  // Generate chat completion using OpenAI's API
  async generateChatCompletion(
    messages: ChatCompletionMessageParam[],
    temperature: number = 0.7,
    model: string = "gpt-4o"
  ): Promise<string> {
    if (!this.apiKey) {
      throw new Error("OpenAI API key is not set");
    }

    try {
      const response = await fetch("https://api.openai.com/v1/chat/completions", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Authorization": `Bearer ${this.apiKey}`
        },
        body: JSON.stringify({
          model,
          messages,
          temperature,
        })
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`OpenAI API error: ${JSON.stringify(error)}`);
      }

      const data = await response.json();
      return data.choices[0].message.content;
    } catch (error) {
      console.error("Error generating chat completion:", error);
      throw error;
    }
  }
}

// Create singleton instance
export const openAIService = new OpenAIService();

File: backend/agents/openai_service.py

import openai
from typing import List, Dict, Any

async def validate_api_key(api_key: str) -> bool:
    """Validate an OpenAI API key by making a simple request."""
    try:
        # Set the API key
        client = openai.OpenAI(api_key=api_key)
        
        # Make a simple request to validate the key
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {"role": "system", "content": "You are a helpful assistant."},
                {"role": "user", "content": "Hello, this is a test."}
            ],
            max_tokens=10
        )
        
        # If we get here without an exception, the key is valid
        return True
    
    except Exception as e:
        print(f"API key validation error: {e}")
        return False

async def generate_embeddings(text: str, api_key: str) -> List[float]:
    """Generate embeddings for the given text using OpenAI's API."""
    try:
        client = openai.OpenAI(api_key=api_key)
        response = client.embeddings.create(
            model="text-embedding-3-small",
            input=text
        )
        return response.data[0].embedding
    
    except Exception as e:
        print(f"Error generating embeddings: {e}")
        raise

async def generate_chat_completion(
    messages: List[Dict[str, str]],
    api_key: str,
    temperature: float = 0.7,
    model: str = "gpt-4o-mini"
) -> str:
    """Generate a completion using OpenAI's Chat Completion API."""
    try:
        client = openai.OpenAI(api_key=api_key)
        response = client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=temperature
        )
        return response.choices[0].message.content
    
    except Exception as e:
        print(f"Error generating chat completion: {e}")
        raise
